src/nlp_expr.jl

```diff
@@ -0,0 +1,871 @@
+
+and the default list of supported multivariate operators is given by:
+
+ * [`MOI.Nonlinear.DEFAULT_MULTIVARIATE_OPERATORS`](@ref)
+
+Additional operators can be registered by setting a [`MOI.UserDefinedFunction`](@ref)
+attribute.
+
+See the full list of operators supported by a [`MOI.ModelLike`](@ref) by
+querying [`MOI.ListOfSupportedNonlinearOperators`](@ref).
+
+## `args`
+
+The vector `args` contains the arguments to the nonlinear function. If the
+operator is univariate, it must contain one element. Otherwise, it may contain
+multiple elements.
+
+Given a subtype of [`AbstractVariableRef`](@ref), `V`, for `GenericNonlinearExpr{V}`,
+each element must be one of the following:
+
+ * A constant value of type `<:Number`
+ * A `V`
+ * A [`GenericAffExpr{C,V}`](@ref)
+ * A [`GenericQuadExpr{C,V}`](@ref)
+ * A [`GenericNonlinearExpr{V}`](@ref)
+
+## Unsupported operators
+
+If the optimizer does not support `head`, an [`MOI.UnsupportedNonlinearOperator`](@ref)
+error will be thrown.
+
+There is no guarantee about when this error will be thrown; it may be thrown
+when the function is first added to the model, or it may be thrown when
+[`optimize!`](@ref) is called.
+
+## Example
+
+To represent the function ``f(x) = sin(x)^2``, do:
+
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x)
+x
+
+julia> f = sin(x)^2
+sin(x) ^ 2.0
+
+julia> f = GenericNonlinearExpr{VariableRef}(
+           :^,
+           GenericNonlinearExpr{VariableRef}(:sin, x),
+           2.0,
+       )
+sin(x) ^ 2.0
+```
+"""
+struct GenericNonlinearExpr{V<:AbstractVariableRef} <: AbstractJuMPScalar
+    head::Symbol
+    args::Vector{Any}
+
+    function GenericNonlinearExpr(head::Symbol, args::Vector{Any})
+        index = findfirst(Base.Fix2(isa, AbstractJuMPScalar), args)
+        if index === nothing
+            error(
+                "Unable to create a nonlinear expression because it did not " *
+                "contain any JuMP scalars. head = $head, args = $args.",
+            )
+        end
+        return new{variable_ref_type(args[index])}(head, args)
+    end
+
+    function GenericNonlinearExpr{V}(
+        head::Symbol,
+        args::Vector{Any},
+    ) where {V<:AbstractVariableRef}
+        return new{V}(head, args)
+    end
+end
+
+"""
+    NonlinearExpr
+
+Alias for `GenericNonlinearExpr{VariableRef}`, the specific
+[`GenericNonlinearExpr`](@ref) used by JuMP.
+"""
+const NonlinearExpr = GenericNonlinearExpr{VariableRef}
+
+variable_ref_type(::GenericNonlinearExpr{V}) where {V} = V
+
+# We include this method so that we can refactor the internal representation of
+# GenericNonlinearExpr without having to rewrite the method overloads.
+function GenericNonlinearExpr{V}(
+    head::Symbol,
+    args...,
+) where {V<:AbstractVariableRef}
+    return GenericNonlinearExpr{V}(head, Any[args...])
+end
+
+Base.length(x::GenericNonlinearExpr) = length(x.args)
+Base.getindex(x::GenericNonlinearExpr, i::Int) = x.args[i]
```

Comment:
That might be confusing, `x` acts as a scalar number so this is the expected behavior:
```julia
julia> 1[1]
1

julia> 1[2]
ERROR: BoundsError: attempt to access Int64 at index [2]
Stacktrace:
 [1] getindex(x::Int64, i::Int64)
   @ Base ./number.jl:98
 [2] top-level scope
   @ REPL[2]:1
```

Reply:
I just removed the methods and their usage